[upd] type aliases (#5676)
JetBrains/kotlin-web-site: docs/topics/type-aliases.md

[//]: # (title: Type aliases)

Type aliases provide alternative names for existing types. They can make long or frequently used type expressions shorter
and easier to understand.

For example, you can create aliases for generic types, function types, and nested or inner classes:

```kotlin
// Generic types
typealias UserIndex = Map<Long, User>
typealias FileTable<K> = MutableMap<K, MutableList<File>>

// Function types
typealias RequestHandler = (Request) -> Response
typealias Predicate<T> = (T) -> Boolean

// Inner and nested classes
class Database {
    inner class Transaction
}
typealias DatabaseTransaction = Database.Transaction
```

A type alias doesn't create a new type. It introduces an alternative name for an existing type. The alias and its underlying
type are interchangeable. For example, when you add `typealias Predicate<T>` and use `Predicate<Int>`, the compiler

> expands it to `(Int) -> Boolean`. You can use a value of the aliased type wherever the underlying type is expected, and

Reviewer: ```suggestion
expands it to `(Int) -> Boolean`. You can use a value declared with the alias wherever the underlying type is expected, and
```

> vice versa:

Reviewer: ```suggestion
the other way around:
```

Reviewer: just to replace the latin with English 👍

```kotlin
typealias Predicate<T> = (T) -> Boolean

fun evaluate(predicate: Predicate<Int>) = predicate(42)

fun main() {
    val isPositive: (Int) -> Boolean = { it > 0 }
    println(evaluate(isPositive))
    // true

    val isValid: Predicate<Int> = { it > 0 }
    println(listOf(1, -2).filter(isValid))
    // [1]
}
```
{kotlin-runnable="true"}

## Declare type aliases

You can declare a type alias:

> * At the top level of a Kotlin file (top-level type aliases).

Reviewer: I don't think the UL adds much here. 🤔  

WDYT about something more concise? 🤔 
You can declare a type alias at the top level of a Kotlin file (a top-level type alias) or inside a class, interface, or object (a nested type alias).

Author: I think helps users clearly see where the subsections come from.

Reviewer: hmm 🤔 if that's the intention, I think it works, but then we should put the focus on that (right now it's in parentheses without links to the relevant sections, so I don't think many people would make this connection)

Author: Yes, it's a good point, updated!

* Inside a class, interface, or object (nested type aliases).

> You can't declare a type alias in a local scope, such as inside a function or lambda expression.

Reviewer: ```suggestion
You can't declare a type alias in a local scope, such as inside a function or [lambda expression](lambdas.md#lambda-expressions-and-anonymous-functions).
```

Reviewer: Since this a bit more advanced concept, let's link to it

The declaration location determines the scope of a type alias, while its visibility determines which code can access it.

> By default, a type alias is `public`. The effective accessibility of a type alias depends on its visibility and the accessibility

Reviewer: ```suggestion
By default, a type alias is `public`. A nested type alias is accessible only where its containing class, interface, or object is accessible.
```

Reviewer: If I understand correctly, this is just for nested aliases. For top-level ones, only their own visibility applies, no? 🤔

Author: No, it applies to both

Reviewer: hmm 🤔 Then I'm confused by this sentence: 

> The effective accessibility of a type alias depends on its visibility and the accessibility of its containing declarations.

What containing declaration do we mean in relation to a top-level declaration?

Author: Sorry, I missunderstood your point! I was talking about underlying type visibility. You are correct, I have updated that part

Reviewer: nice that's a great improvement 🎉 — my only concern is the "effective accessibility" part now. The docs don't use this term, because it can be a bit vague - so I still prefer a more exact wording, like:

> A nested type alias is accessible only where its containing class, interface, or object is accessible. 

this sentence sidesteps the issue with the abstract noun phrase/technical term

of its containing declarations. For example, a `public` alias inside an `internal` class isn't accessible from outside

> the module because the containing class isn't accessible there. When you declare a type alias, its underlying type must

Reviewer: ```suggestion
the module because the containing class isn't accessible there. 

When you declare a type alias, its underlying type must
```

Reviewer: must be at least as visible sounds a bit mechancial 🤔
WDYT about having a simpler sentence — the example you use below demonstrates this beautifully we could even cut the last sentence:

A type alias can't expose a type with more restrictive [visibility](visibility-modifiers.md) than its own. 
For example, ...

> the module because the containing class isn't accessible there. When you declare a type alias, its underlying type must

Reviewer: ```suggestion
the module. When you declare a type alias, its underlying type must
```

be at least as visible as the alias. A type alias can't expose a type whose visibility is more restrictive than the alias.

> For example, a `public` type alias can't refer to a `private` class. Learn how visibility affects access in [](visibility-modifiers.md).

Reviewer: ```suggestion
For example, a `public` type alias can't refer to a `private` class.
```

### Top-level type aliases

A top-level type alias is a package-level declaration. Within the same package, you can refer to an alias by its unqualified name.
To use the alias from another package, import the alias or refer to it by its qualified name:

```kotlin

> // user-id.kt

Reviewer: ```suggestion
// UserId.kt
```

Reviewer: Just to follow [Kotlin naming conventions](https://kotlinlang.org/docs/coding-conventions.html#source-file-names)

package org.example.users

typealias UserId = Long

> // user-service.kt

Reviewer: ```suggestion
// UserService.kt
```

package org.example.services

> import org.example.users.UserId

Reviewer: could we add an unqualified name here? 🙏 

something like (but up to you )

fun findUser(id: UserId) { // ... 
}

}
```

### Nested type aliases

Declare a type alias inside a class, interface, or object when the alternative name is relevant only in the context of
that declaration. This keeps the alias close to the code that uses it and avoids adding another name to the package scope.

> Nested type aliases allow for cleaner, more maintainable code by improving encapsulation, reducing package-level clutter,

Reviewer: ```suggestion

Nested type aliases allow for cleaner, more maintainable code by improving encapsulation, reducing package-level clutter,
```

Reviewer: hmm 🤔 
I think here we should focus on the rules first WDYT?

We could put this sentence about their use case to the top of the section: 
> Nested type aliases allow for cleaner, more maintainable code by improving encapsulation, reducing package-level clutter, and simplifying internal implementations.

Then we could keep this separately here.

Nested type aliases follow the same scope and name resolution rules as [nested classes](link).

and simplifying internal implementations. Nested type aliases follow the same scope and name-resolution rules as [nested classes](nested-classes.md).

Within the containing declaration, you can refer to the alias by its unqualified name. Outside the declaration, qualify
the alias with the name of its containing declaration:

```kotlin
class UserRepository {
    typealias UserIndex = Map<UserId, User>

> 

Reviewer: ```suggestion

    // Refers to the alias by its unqualified name inside UserRepository
```

Reviewer: (suggestion)
perhaps overkill, but I thought it could help up to you

Author: No, I like it! Thanks :)

    fun saveAll(users: UserIndex) {
        // ...
    }
}

> 

Reviewer: ```suggestion

// Refers to the alias by its qualified name outside UserRepository
```

fun synchronizeUsers(users: UserRepository.UserIndex) {
  // ...
}
```

> > Nested type aliases are not supported in Kotlin Multiplatform [`expect/actual` declarations](https://kotlinlang.org/docs/multiplatform/multiplatform-expect-actual.html).

Reviewer: ```suggestion
> Nested type aliases aren't supported in Kotlin Multiplatform [`expect/actual` declarations](https://kotlinlang.org/docs/multiplatform/multiplatform-expect-actual.html).
```

>
{style="note"}

#### Type parameters

> A nested type alias can't capture type parameters from its containing declaration. Capturing occurs when the alias refers

Reviewer: As a general comment. It feels strange to lead with what the user can't do 🤔 

What do you think about flipping this section a bit? 

To use type parameters in a nested type alias, add them to the alias declaration:

[correct example]

In this example, `Path` declares `T`.  In `Graph.Path<String>`, `String` provides the type argument for `T` and doesn't depend on the `Node` type parameter declared by `Graph`. (a bit wordy, but I couldn't come up with a smoother version. 😢 )

If you refer to a type parameter declared by the class or interface that contains the alias, the compiler reports an error:

[incorrect example]

Personally, I also would suggest dropping the part about "capturing" — I see that the original had it, but I think we can make this more understandable without it.

to a type parameter declared by its outer class or interface:

```kotlin
class Graph<Node> {

>     // Error: Path captures Graph's Node type parameter

Reviewer: this is a bit misleading — it seems like this is the error compiler would give I'd suggest:

This results in an error, because Path captures Graph's Node type parameter

Then add below: 
// Unresolved reference 'Node'.

    typealias Path = List<Node>
}
```

In this example, `Path` can't use `Node` because a nested type alias doesn't capture type parameters from its containing class.

> To make the alias generic, declare its own type parameter:

Reviewer: ```suggestion

To make the alias generic, declare its own type parameter:
```

expands it to `(Int) -> Boolean`. You can use a value of the aliased type wherever the underlying type is expected, and
vice versa:

```kotlin
typealias Predicate<T> = (T) -> Boolean

fun evaluate(predicate: Predicate<Int>) = predicate(42)

fun main() {
    val isPositive: (Int) -> Boolean = { it > 0 }
    println(evaluate(isPositive))
    // true

    val isValid: Predicate<Int> = { it > 0 }
    println(listOf(1, -2).filter(isValid))
    // [1]
}
```
{kotlin-runnable="true"}

## Declare type aliases

You can declare a type alias:

* At the top level of a Kotlin file (top-level type aliases).
* Inside a class, interface, or object (nested type aliases).

You can't declare a type alias in a local scope, such as inside a function or lambda expression.

### Top-level type aliases

A top-level type alias is a package-level declaration. Within the same package, you can refer to an alias by its unqualified name.
To use the alias from another package, import the alias or refer to it by its qualified name:

```kotlin
// user-id.kt
package org.example.users

typealias UserId = Long

// user-service.kt
package org.example.services

import org.example.users.UserId
}
```

> The package determines whether you can refer to the alias by its unqualified name or need to import it or use its fully

Reviewer 2: I do not understand what is the meaning of this first sentence. What new idea does it introduces compared to the text before the snippet?

Author: removed

```kotlin
private typealias ResponseCache = MutableMap<String, CachedResponse>
```

Since top-level type aliases are not class members, you can't mark them as `protected`.

### Nested type aliases

Declare a type alias inside a class, interface, or object when the alternative name is relevant only in the context of
that declaration. This keeps the alias close to the code that uses it and avoids adding another name to the package scope.
Nested type aliases allow for cleaner, more maintainable code by improving encapsulation, reducing package-level clutter,
and simplifying internal implementations. Nested type aliases follow the same scope and name-resolution rules as [nested classes](nested-classes.md).

Within the containing declaration, you can refer to the alias by its unqualified name. Outside the declaration, qualify
the alias with the name of its containing declaration:

```kotlin
class UserRepository {
    typealias UserIndex = Map<UserId, User>

    fun saveAll(users: UserIndex) {
        // ...
    }
}

fun synchronizeUsers(users: UserRepository.UserIndex) {
  // ...
}
```

> The containing declaration establishes the alias's scope, but visibility can restrict access further. Therefore,

Reviewer 2: There are two places with the description of the fact that type aliases could have visibility modifier. Probably it is better to just mention it in some shared section (first one or "Declare type aliases"). Also I do not think that such a detailed explanation of how visibility works and what kinds of visibilities are there is needed. It works the same way as for any other entity.

Author: Fair enough, removed description of visibility modifiers

the effective accessibility of a nested alias depends on both the alias and its containing declaration. For example, you can't access
a `public` alias inside an `internal` class from outside the module because the containing class isn't accessible there.

> A type alias also can't expose an underlying type whose visibility is more restrictive than the alias. For example, a `public`

Reviewer 2: This applies not only to nested type aliases, but to both of them. Probably it should be in any of the common sections

Author: Moved visability info to the top section and proveded relevant link to avoid additional explanation

```kotlin
// Generic types
typealias UserIndex = Map<Long, User>
typealias FileTable<K> = MutableMap<K, MutableList<File>>

// Function types
typealias RequestHandler = (Request) -> Response
typealias Predicate<T> = (T) -> Boolean

// Inner and nested classes
class Database {
    inner class Transaction
}
typealias DatabaseTransaction = Database.Transaction
```

A type alias doesn't create a new type. It introduces an alternative name for an existing type. The alias and its underlying
type are interchangeable. For example, when you add `typealias Predicate<T>` and use `Predicate<Int>`, the compiler
expands it to `(Int) -> Boolean`. You can use a value declared with the alias wherever the underlying type is expected, and
the other way around:

```kotlin
typealias Predicate<T> = (T) -> Boolean

fun evaluate(predicate: Predicate<Int>) = predicate(42)

fun main() {
    val isPositive: (Int) -> Boolean = { it > 0 }
    println(evaluate(isPositive))
    // true

    val isValid: Predicate<Int> = { it > 0 }
    println(listOf(1, -2).filter(isValid))
    // [1]
}
```
{kotlin-runnable="true"}

## Declare type aliases

You can declare a type alias:

> * At the top level of a Kotlin file with [top-level type aliases](#top-level-type-aliases).

Reviewer: ```suggestion
* At the top level of a Kotlin file, as a [top-level type alias](#top-level-type-aliases).
```

> * Inside a class, interface, or object with [nested type aliases](#nested-type-aliases).

Reviewer: ```suggestion
* Inside a class, interface, or object, as a [nested type alias](#nested-type-aliases).
```

Reviewer: as is a bit cleaner than with in this case imho. 🤔 with suggests to me that you use a type alias to declare something else.

You can't declare a type alias in a local scope, such as inside a function or [lambda expression](lambdas.md#lambda-expressions-and-anonymous-functions).

The declaration location determines the scope of a type alias, while its visibility determines which code can access it.
By default, a type alias is `public`. For [nested type aliases](#nested-type-aliases), effective accessibility also depends on the
accessibility of the containing declaration. For example, a `public` alias inside an `internal` class isn't accessible
from outside the module.

A type alias can't expose an underlying type with more restrictive [visibility](visibility-modifiers.md) than its own. For example, a `public`
type alias can't refer to a `private` class.

### Top-level type aliases

A top-level type alias is a package-level declaration. Within the same package, you can refer to an alias by its unqualified name.
To use the alias from another package, import the alias or refer to it by its qualified name:

```kotlin
// UserId.kt
package org.example.users

typealias UserId = Long

> 

Reviewer: some nits here:
Technically it would be more correct imho to:
- place the comment ABOVE the line (fun createUser(id: UserId) { and similar here)) instead of inside
- let's use third person comments and not passive

Something like
// Uses the unqualified name within the same package / or // Refers to the alias within the same package by its unqualified name
// Uses the imported alias by its unqualified name / // ...
// Uses the fully qualified name / // ...

Declare a type alias inside a class, interface, or object when the alternative name is relevant only in the context of
that declaration. This keeps the alias close to the code that uses it and avoids adding another name to the package scope.

Within the containing declaration, you can refer to the alias by its unqualified name. Outside the declaration, qualify
the alias with the name of its containing declaration:

```kotlin
class UserRepository {
    typealias UserIndex = Map<UserId, User>

    // Refers to the alias by its unqualified name inside UserRepository
    fun saveAll(users: UserIndex) {
        // ...
    }
}

// Refers to the alias by its qualified name outside UserRepository
fun synchronizeUsers(users: UserRepository.UserIndex) {
  // ...
}
```

> Nested type aliases aren't supported in Kotlin Multiplatform [`expect/actual` declarations](https://kotlinlang.org/docs/multiplatform/multiplatform-expect-actual.html).
>
{style="note"}

#### Type parameters

To use type parameters in a nested type alias, add them to the alias declaration:

> ```

Reviewer: ```suggestion
```kotlin
```

class Graph<Node> {
    typealias Path<T> = List<T>
}

val cityPath: Graph.Path<String> = listOf("London", "Berlin")
```

In this example, `Path` declares its type parameter `T`. In `Graph.Path<String>`, `String` is the type argument for `T`
and is independent of the `Node` type parameter declared by `Graph`.

If you refer to a type parameter declared by its containing class or interface, the compiler reports an error:

> ```

Reviewer: ```suggestion
```kotlin
```

The declaration location determines the scope of a type alias, while its visibility determines which code can access it.
By default, a type alias is `public`. A [nested type alias](#nested-type-aliases) is accessible only where its containing class, interface,
or object is accessible. For example, a `public` alias inside an `internal` class isn't accessible from outside the module.

A type alias can't expose an underlying type with more restrictive [visibility](visibility-modifiers.md) than its own. For example, a `public`
type alias can't refer to a `private` class.

### Top-level type aliases

A top-level type alias is a package-level declaration. Within the same package, you can refer to an alias by its unqualified name.
To use the alias from another package, import the alias or refer to it by its qualified name:

```kotlin
// UserId.kt
package org.example.users

typealias UserId = Long

// Refers to the alias within the same package by its unqualified name
fun createUser(id: UserId) {
    // ...
}

// UserService.kt
package org.example.services

import org.example.users.UserId

> // Uses the imported alias by its unqualified name 

Reviewer: ```suggestion
// Uses the imported alias by its unqualified name
```

fun findUser(id: UserId) {
    // ...
}

// Uses the fully qualified name
fun deleteUser(id: org.example.users.UserId) {
    // ...
}
```

### Nested type aliases

Nested type aliases allow for cleaner, more maintainable code by improving encapsulation, reducing package-level clutter,
and simplifying internal implementations. Nested type aliases follow the same scope and name-resolution rules as [nested classes](nested-classes.md).

Declare a type alias inside a class, interface, or object when the alternative name is relevant only in the context of
that declaration. This keeps the alias close to the code that uses it and avoids adding another name to the package scope.

Within the containing declaration, you can refer to the alias by its unqualified name. Outside the declaration, qualify
the alias with the name of its containing declaration:

```kotlin
class UserRepository {
    typealias UserIndex = Map<UserId, User>

    // Refers to the alias by its unqualified name inside UserRepository
    fun saveAll(users: UserIndex) {
        // ...
    }
}

// Refers to the alias by its qualified name outside UserRepository
fun synchronizeUsers(users: UserRepository.UserIndex) {

>   // ...

Reviewer: ```suggestion
    // ...
```

You can't declare a type alias in a local scope, such as inside a function or lambda expression.

The declaration location determines the scope of a type alias, while its visibility determines which code can access it.
By default, a type alias is `public`. The effective accessibility of a type alias also depends on the accessibility of

> its containing declarations and visibility of the underlying type. A type alias can't expose a type whose visibility is

Reviewer 2: > The effective accessibility of a type alias also depends on the accessibility of its containing declarations and visibility of the underlying type

Effective accessibility of a type alias does not depends on the visibility of the underlying type. The reason for that is in your next sentence: "A type alias can't expose a type whose visibility is more restrictive than the alias". So, the correctness of the visibility of underlying type is checked in regards to typealias' visibility, but it does not affect the visibility furthermore
